src/H5warnings.h

```diff
@@ -12,62 +12,315 @@
+ *   consider shutting down the warning flag globally using the compiler
+ *   flags (like -Wno-foo-warning).
+ *
+ * - The macro should have a helpful comment about why the warning should
+ *   be suppressed and not corrected.
+ *
+ * - The macro name should reflect the actual problem, not the name of the
+ *   compiler warning option. Names are of the form H5_WARN_<THING>_(ON|OFF).
+ *
+ * - Be careful with the ifdefs. clang defines __GNUC__, for example, so
+ *   you can't simply check for that if you have a gcc-specific warning.
+ *   Check for compiler version numbers in the macros to avoid warnings
+ *   about undefined diagnostics in older compilers.
+ *
+ * - Add the new macro to the list in the .clang-format file.
  */
 
 #ifndef H5warnings_H
 #define H5warnings_H
 
-/* Macros for enabling/disabling particular GCC / clang warnings
+/* Macros for enabling/disabling particular gcc / clang warnings
  *
  * (see the following web-sites for more info:
  *      http://www.dbp-consulting.com/tutorials/SuppressingGCCWarnings.html
  *      http://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html#Diagnostic-Pragmas
+ *
+ * _Pragma is C99 and should work with all compilers, though permitted
+ * placement may vary
+ *
+ * "GCC" in the pragma works for both clang/llvm and gcc
  */
-#define H5_DIAG_JOINSTR(x, y) x y
-#define H5_DIAG_DO_PRAGMA(x)  _Pragma(#x)
-#define H5_DIAG_PRAGMA(x)     H5_DIAG_DO_PRAGMA(GCC diagnostic x)
+#define H5_WARN_JOINSTR(x, y) x y
+#define H5_WARN_DO_PRAGMA(x)  _Pragma(#x)
+#define H5_WARN_PRAGMA(x)     H5_WARN_DO_PRAGMA(GCC diagnostic x)
 
 /* Allow suppression of compiler diagnostics unless H5_SHOW_ALL_WARNINGS is
- *      defined (enabled with '--enable-show-all-warnings' configure option).
+ * defined
+ */
+/* clang-format off */
+#ifdef H5_SHOW_ALL_WARNINGS
+    #define H5_WARN_OFF(x)
+    #define H5_WARN_ON(x)
+#else
+    #if (((__GNUC__ * 100) + __GNUC_MINOR__) >= 406) || defined(__clang__)
+        /* gcc 4.6+ /clang */
+        #define H5_WARN_OFF(x) H5_WARN_PRAGMA(push) H5_WARN_PRAGMA(ignored H5_WARN_JOINSTR("-W", x))
+        #define H5_WARN_ON(x)  H5_WARN_PRAGMA(pop)
+    #endif
+#endif
+/* clang-format on */
+
+/*********************
+ * SPECIFIC WARNINGS *
+ *********************/
+
+/* Suppress warnings about C11 extensions */
+#if defined(__clang__)
+#define H5_WARN_C11_EXTENSIONS_OFF H5_WARN_OFF("c11-extensions")
+#define H5_WARN_C11_EXTENSIONS_ON  H5_WARN_ON("c11-extensions")
+#else
+#define H5_WARN_C11_EXTENSIONS_OFF
+#define H5_WARN_C11_EXTENSIONS_ON
+#endif
+
+/* Suppress warnings about bad cast alignment. These should be corrected,
+ * not suppressed in the main library, but might appear in test code.
  */
-#ifndef H5_SHOW_ALL_WARNINGS
-#define H5_DIAG_OFF(x) H5_DIAG_PRAGMA(push) H5_DIAG_PRAGMA(ignored H5_DIAG_JOINSTR("-W", x))
-#define H5_DIAG_ON(x)  H5_DIAG_PRAGMA(pop)
+#if defined(__clang__) || defined(__GNUC__)
+#define H5_WARN_CAST_ALIGNMENT_OFF H5_WARN_OFF("cast-align")
```

Comment:
Out of all the warning types, `cast-align` is one that I think we should never allow suppressing warnings for. While it's fine on some platforms it's certainly not on others and will likely lead to subtle issues.